File: android/supabase/src/main/kotlin/io/github/notkanaa/equipe/supabase/RestClient.kt
package io.github.notkanaa.equipe.supabase

import io.github.notkanaa.equipe.core.AppError
import io.ktor.client.HttpClient
import io.ktor.client.request.header
import io.ktor.client.request.request
import io.ktor.client.request.setBody
import io.ktor.client.statement.bodyAsBytes
import io.ktor.content.ByteArrayContent
import io.ktor.http.ContentType
import io.ktor.http.HttpHeaders
import io.ktor.http.HttpMethod
import kotlinx.coroutines.currentCoroutineContext
import kotlinx.coroutines.ensureActive
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject
import java.util.UUID
import kotlin.coroutines.cancellation.CancellationException

/** The signed-in user of a client session and a valid access token. */
internal data class Credentials(val userId: UUID, val accessToken: String) {
    override fun toString(): String = "Credentials(userId=$userId, accessToken=…)"
}

/** Provides the credentials of the current session; throws [AppError.NotAuthenticated] when signed out. */
internal interface CredentialsProvider {
    suspend fun credentials(): Credentials

    /**
     * The server refused [rejected] although its local expiry had not passed (signing key rotated, device clock behind,
     * account deleted): refreshes the session and returns the new credentials. When the session cannot be refreshed
     * (revoked, account deleted), the session is removed (the client signs out) and [AppError.NotAuthenticated] is
     * thrown; offline, [AppError.Network].
     */
    suspend fun refreshedCredentials(rejected: Credentials): Credentials
}

/** One PostgREST answer, before error mapping. */
internal class RestResponse(val status: Int, val body: ByteArray) {
    val isSuccess: Boolean get() = status in 200..299

    /** The mapped error of a non-2xx answer, null for a success. */
    val error: AppError? by lazy { if (isSuccess) null else SupabaseErrorMapping.postgrest(status, body) }

    /**
     * The server refused the session of the request: HTTP 401 (token refused by PostgREST) or `not_authenticated`
     * (token accepted, but its account no longer exists). Nothing was written.
     */
    val refusesTheSession: Boolean get() = error == AppError.NotAuthenticated

    /** `not_authenticated` raised by our SQL with an accepted token: the account of the token no longer exists. */
    val accountIsGone: Boolean get() = status != 401 && refusesTheSession

    /** The body of a 2xx answer; the mapped error otherwise. */
    fun data(): ByteArray {
        error?.let { throw it }
        return body
    }
}

/**
 * Minimal PostgREST client (port of RestClient.swift): the adapters build every request themselves ([RestQuery]) so
 * that the query strings are exactly those of docs/CONTRACTS.md §4.3 and the HTTP status of an error reaches
 * [io.github.notkanaa.equipe.core.BackendErrorMapper] (401 vs 403 matters for `42501`).
 *
 * @param projectUrl `<project>` (`/rest/v1` is appended).
 * @param http a Ktor client without default request settings (tests plug a mock engine).
 */
internal class RestClient(
    projectUrl: String,
    private val apiKey: String,
    private val http: HttpClient,
    val session: CredentialsProvider,
) {
    /** `<project>/rest/v1`. */
    val restUrl: String = projectUrl.trimEnd('/') + "/rest/v1"

    /** The current session's credentials ([AppError.NotAuthenticated] when there is none). */
    suspend fun credentials(): Credentials = session.credentials()

    /**
     * Sends the request built for the current session and returns the body of a 2xx answer.
     *
     * When the server refuses the session ([RestResponse.refusesTheSession]), the session is refreshed once and the
     * request sent again with the new token: a token refused before its local expiry (signing key rotated, clock
     * behind) is replaced, and a dead session (revoked, account deleted) fails to refresh, which signs the device out
     * instead of leaving the app « signed in » with every call failing until the token expires.
     */
    suspend fun send(build: (Credentials) -> RestRequest): ByteArray {
        val credentials = credentials()
        val first = response(build(credentials), credentials)
        if (!first.refusesTheSession) return first.data()
        val fresh = session.refreshedCredentials(credentials)
        return response(build(fresh), fresh).data()
    }

    /** Sends [request] once and returns the raw answer; only transport failures throw. */
    suspend fun response(request: RestRequest, credentials: Credentials): RestResponse {
        val url = request.url(restUrl)
        return try {
            val response = http.request(url) {
                method = when (request.method) {
                    RestRequest.Method.GET -> HttpMethod.Get
                    RestRequest.Method.POST -> HttpMethod.Post
                    RestRequest.Method.PATCH -> HttpMethod.Patch
                }
                header("apikey", apiKey)
                header(HttpHeaders.Authorization, "Bearer ${credentials.accessToken}")
                header(HttpHeaders.Accept, "application/json")
                request.prefer?.let { header("Prefer", it) }
                request.encodedBody()?.let {
                    setBody(ByteArrayContent(it.toByteArray(Charsets.UTF_8), ContentType.Application.Json))
                }
            }
            RestResponse(response.status.value, response.bodyAsBytes())
        } catch (error: CancellationException) {
            throw error
        } catch (error: Throwable) {
            currentCoroutineContext().ensureActive()
            throw SupabaseErrorMapping.transport(error)
        }
    }

    /** Sends and decodes the JSON answer with [decode] (any decoding failure is an unexpected answer). */
    suspend fun <T> fetch(decode: (JsonElement) -> T, build: (Credentials) -> RestRequest): T =
        RestDecoding.decode(send(build), decode)

    /** Sends and decodes a JSON array of rows, all of them (an unknown enum value fails the answer). */
    suspend fun <T> fetchAll(decodeRow: (JsonObject) -> T, build: (Credentials) -> RestRequest): List<T> =
        fetch({ element ->
            val array = element as? JsonArray ?: throw MalformedAnswer("expected an array")
            array.map { decodeRow(it.asObject("row")) }
        }, build)

    /**
     * Sends and decodes a JSON array, leaving out the rows with a value this client does not know
     * ([RestDecoding.decodeRows]): a newer server must not make a whole list unreadable.
     */
    suspend fun <T> fetchRows(decodeRow: (JsonObject) -> T, build: (Credentials) -> RestRequest): List<T> =
        RestDecoding.decodeRows(send(build), decodeRow)
}

File: Packages/TeamTasksKit/Sources/TeamTasksSupabase/Rest/RestClient.swift
import Foundation
import TeamTasksCore

#if canImport(FoundationNetworking)
import FoundationNetworking
#endif

/// Sends HTTP requests (injectable, so unit tests can answer with JSON fixtures).
protocol HTTPTransport: Sendable {
    func send(_ request: URLRequest) async throws -> (Data, HTTPURLResponse)
}

struct URLSessionTransport: HTTPTransport {
    let session: URLSession

    func send(_ request: URLRequest) async throws -> (Data, HTTPURLResponse) {
        let (data, response) = try await session.data(for: request)
        guard let http = response as? HTTPURLResponse else { throw URLError(.badServerResponse) }
        return (data, http)
    }
}

/// The signed-in user of a client session and a valid access token.
struct Credentials: Sendable, Hashable {
    let userId: UUID
    let accessToken: String
}

/// Provides the credentials of the current session; throws `AppError.notAuthenticated` when signed out.
protocol CredentialsProvider: Sendable {
    func credentials() async throws -> Credentials
    /// The server refused `rejected` although its local expiry had not passed (signing key rotated, device clock
    /// behind, account deleted): refreshes the session and returns the new credentials. When the session cannot be
    /// refreshed (revoked, account deleted), the session is removed (the Auth client emits `.signedOut`) and
    /// `.notAuthenticated` is thrown; offline, `.network`.
    func refreshedCredentials(after rejected: Credentials) async throws -> Credentials
}

/// One PostgREST answer, before error mapping.
struct RestResponse: Sendable {
    let status: Int
    let body: Data

    var isSuccess: Bool { (200..<300).contains(status) }

    /// The mapped error of a non-2xx answer, nil for a success.
    var error: AppError? {
        isSuccess ? nil : SupabaseErrorMapping.postgrest(status: status, body: body)
    }

    /// The server refused the session of the request: HTTP 401 (token refused by PostgREST) or
    /// `not_authenticated` (token accepted, but its account no longer exists). Nothing was written.
    var refusesTheSession: Bool { error == .notAuthenticated }

    /// `not_authenticated` raised by our SQL (`private.require_uid`) with an accepted token: the account of the
    /// token no longer exists.
    var accountIsGone: Bool { status != 401 && refusesTheSession }

    /// The body of a 2xx answer; the mapped error otherwise.
    func data() throws -> Data {
        if let error { throw error }
        return body
    }
}

/// Minimal PostgREST client: the adapters build every request themselves (`RestQuery`) so that the query
/// strings are exactly those of docs/CONTRACTS.md §4.3 and the HTTP status of an error reaches
/// `BackendErrorMapper` (supabase-swift's `PostgrestError` drops it, and 401 vs 403 matters for `42501`).
struct RestClient: Sendable {
    /// `<project>/rest/v1`.
    let restURL: URL
    let apiKey: String
    let transport: any HTTPTransport
    let session: any CredentialsProvider

    init(projectURL: URL, apiKey: String, transport: any HTTPTransport, session: any CredentialsProvider) {
        restURL = projectURL.appendingPathComponent("rest").appendingPathComponent("v1")
        self.apiKey = apiKey
        self.transport = transport
        self.session = session
    }

    /// The current session's credentials (`.notAuthenticated` when there is none).
    func credentials() async throws -> Credentials {
        try await session.credentials()
    }

    /// Sends `request` for the current session and returns the body of a 2xx answer.
    ///
    /// When the server refuses the session (`RestResponse.refusesTheSession`), the session is refreshed once and
    /// the request sent again with the new token: a token refused before its local expiry (signing key rotated,
    /// clock behind) is replaced, and a dead session (revoked, account deleted) fails to refresh, which signs the
    /// device out instead of leaving the app « signed in » with every call failing until the token expires.
    func send(_ build: (Credentials) -> RestRequest) async throws -> Data {
        let credentials = try await credentials()
        let first = try await response(to: build(credentials), credentials: credentials)
        guard first.refusesTheSession else { return try first.data() }
        let fresh = try await session.refreshedCredentials(after: credentials)
        return try await response(to: build(fresh), credentials: fresh).data()
    }

    /// Sends `request` once and returns the body of a 2xx answer (no refresh).
    func send(_ request: RestRequest, credentials: Credentials) async throws -> Data {
        try await response(to: request, credentials: credentials).data()
    }

    /// Sends `request` once and returns the raw answer; only transport failures throw.
    func response(to request: RestRequest, credentials: Credentials) async throws -> RestResponse {
        guard let url = request.url(restURL: restURL) else { throw AppError.misconfigured }
        var urlRequest = URLRequest(url: url)
        urlRequest.httpMethod = request.method.rawValue
        urlRequest.setValue(apiKey, forHTTPHeaderField: "apikey")
        urlRequest.setValue("Bearer \(credentials.accessToken)", forHTTPHeaderField: "Authorization")
        urlRequest.setValue("application/json", forHTTPHeaderField: "Accept")
        if let prefer = request.prefer {
            urlRequest.setValue(prefer, forHTTPHeaderField: "Prefer")
        }
        if let body = request.body {
            urlRequest.setValue("application/json", forHTTPHeaderField: "Content-Type")
            urlRequest.httpBody = try body.encoded()
        }

        do {
            let (data, response) = try await transport.send(urlRequest)
            return RestResponse(status: response.statusCode, body: data)
        } catch {
            if Task.isCancelled { throw CancellationError() }
            throw SupabaseErrorMapping.transport(error)
        }
    }

    /// Sends and decodes the JSON answer.
    func fetch<Value: Decodable>(_ type: Value.Type, _ build: (Credentials) -> RestRequest) async throws -> Value {
        try RestDecoding.decode(type, from: try await send(build))
    }

    /// Sends and decodes a JSON array, leaving out the rows with a value this client does not know
    /// (`LossyRows`): a newer server must not make a whole list unreadable.
    func fetchRows<Row: Decodable>(_ type: Row.Type, _ build: (Credentials) -> RestRequest) async throws -> [Row] {
        try RestDecoding.decode(LossyRows<Row>.self, from: try await send(build)).rows
    }
}

/// JSON decoding of PostgREST answers: timestamps with 0 to 6 fractional digits (`PostgresTimestamp`).
enum RestDecoding {
    static func makeDecoder() -> JSONDecoder {
        let decoder = JSONDecoder()
        decoder.dateDecodingStrategy = .custom { decoder in
            let container = try decoder.singleValueContainer()
            let text = try container.decode(String.self)
            guard let date = PostgresTimestamp.parse(text) else {
                throw DecodingError.dataCorruptedError(in: container, debugDescription: "Invalid timestamp \(text)")
            }
            return date
        }
        return decoder
    }

    static func decode<Value: Decodable>(_ type: Value.Type, from data: Data) throws -> Value {
        do {
            return try makeDecoder().decode(type, from: data)
        } catch {
            throw AppError.unknown(SupabaseErrorMapping.unexpectedAnswer)
        }
    }
}

// MARK: - Forward compatibility

/// A value of a server enum this client does not know (added by a later migration).
struct UnknownEnumValue: Error, Sendable, Hashable {
    let type: String
    let value: String
}

/// An enum column (`task_status`, `task_priority`, `member_role`): known values decode as usual, an unknown one
/// throws `UnknownEnumValue` (not a `DecodingError`), so that `LossyRows` leaves only that row out. Anywhere else
/// (a single row, an RPC result) it fails the decoding like any malformed answer.
struct Known<Value: RawRepresentable & Sendable & Hashable>: Decodable, Sendable, Hashable where Value.RawValue == String {
    let value: Value

    init(_ value: Value) {
        self.value = value
    }

    init(from decoder: any Decoder) throws {
        let raw = try decoder.singleValueContainer().decode(String.self)
        guard let value = Value(rawValue: raw) else {
            throw UnknownEnumValue(type: String(describing: Value.self), value: raw)
        }
        self.value = value
    }
}

/// A JSON array whose rows holding an unknown enum value are left out (they cannot be shown, nor safely edited:
/// an edit would overwrite the unknown value); any other malformed row fails the whole array.
struct LossyRows<Row: Decodable>: Decodable {
    let rows: [Row]
    /// Number of rows left out.
    let skipped: Int

    init(from decoder: any Decoder) throws {
        var container = try decoder.unkeyedContainer()
        var rows: [Row] = []
        var skipped = 0
        if let count = container.count {
            rows.reserveCapacity(count)
        }
        while !container.isAtEnd {
            do {
                rows.append(try container.decode(Row.self))
            } catch is UnknownEnumValue {
                // A failed element does not advance the container: step over it.
                _ = try container.decode(SkippedRow.self)
                skipped += 1
            }
        }
        self.rows = rows
        self.skipped = skipped
    }

    private struct SkippedRow: Decodable {
        init(from decoder: any Decoder) throws {}
    }
}
